compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,522 @@
+    }
+    cost
+}
+
+/// Identifies all simple conditionals in the function and flattens them
+fn flatten_function(function: &mut Function, no_predicates: &mut HashMap<FunctionId, bool>) {
+    // This pass is dedicated to brillig functions
+    if !function.runtime().is_brillig() {
+        return;
+    }
+    let cfg = ControlFlowGraph::with_function(function);
+    let mut stack = vec![function.entry_block()];
+    let mut processed = HashSet::new();
+    let mut conditionals = Vec::new();
+
+    // 1. Process all blocks of the cfg, starting from the root and following the successors
+    while let Some(block) = stack.pop() {
+        // Avoid cycles
+        if processed.contains(&block) {
+            continue;
+        }
+        processed.insert(block);
+
+        // Identify the simple conditionals
+        if let Some(conditional) = is_conditional(block, &cfg, function) {
+            // no need to check the branches, process the join block directly
+            stack.push(conditional.block_exit);
+            conditionals.push(conditional);
+        } else {
+            stack.extend(cfg.successors(block));
+        }
+    }
+
+    // 2. Flatten all simple conditionals
+    // process basic conditionals in reverse order so that
+    // a conditional does not impact the previous ones
+    conditionals.reverse();
+    Context::flatten_multiple(&conditionals, function, no_predicates);
+}
+
+impl<'f> Context<'f> {
+    fn flatten_single_conditional(
+        &mut self,
+        conditional: &BasicConditional,
+        no_predicates: &mut HashMap<FunctionId, bool>,
+    ) {
+        // Manually inline 'then', 'else' and 'exit' into the entry block
+        //0. initialize the context for flattening a 'single conditional'
+        let mut queue = vec![];
+        self.target_block = conditional.block_entry;
+        self.no_predicate = true;
+        //1. process 'then' branch
+        self.inline_block(conditional.block_entry, no_predicates);
+        let to_process = self.handle_terminator(conditional.block_entry, &queue);
+        queue.extend(to_process);
+        if let Some(then) = conditional.block_then {
+            assert_eq!(queue.pop(), conditional.block_then);
+            self.inline_block(then, no_predicates);
+            let to_process = self.handle_terminator(then, &queue);
+
+            for incoming_block in to_process {
+                if !queue.contains(&incoming_block) {
+                    queue.push(incoming_block);
+                }
+            }
+        }
+
+        //2. process 'else' branch, in case there is no 'then'
+        let next = queue.pop();
+        if next == conditional.block_else {
+            let next = next.unwrap();
+            self.inline_block(next, no_predicates);
+            let to_process = self.handle_terminator(next, &queue);
+            for incoming_block in to_process {
+                if !queue.contains(&incoming_block) {
+                    queue.push(incoming_block);
```

Comment:
I don't see see these values being popped: the 3rd section handles the exit block, and there is no looping 👀

Reply:
Same as the previous conversation, I have removed this code that is indeed not needed here.

Reply:
I think I forgot to push the commit, should be ok now.